@planetarium/account-web3-secret-storage/src/Web3KeyStore.ts

```diff
@@ -0,0 +1,224 @@
+import { KeyId } from "./KeyId.js";
+import { PassphraseEntry } from "./PassphraseEntry.js";
+import {
+  type AccountDeletion,
+  type AccountGeneration,
+  type AccountImportation,
+  type AccountMetadata,
+  type AccountRetrieval,
+  Address,
+  type ImportableKeyStore,
+  RawPrivateKey,
+} from "@planetarium/account";
+import {
+  type KeystoreAccount,
+  decryptKeystoreJson,
+  encryptKeystoreJson,
+  isKeystoreJson,
+} from "ethers";
+import { Dirent } from "node:fs";
+import * as fs from "node:fs/promises";
+import { homedir } from "node:os";
+import * as path from "node:path";
+
+export interface Web3KeyStoreOptions {
+  path?: string;
+  passphraseEntry: PassphraseEntry;
+}
+
+/**
+ * Determines the default key store path.  It depends on the platform:
+ *
+ * - Linux/macOS: `$HOME/.config/planetarium/keystore`
+ * - Windows: `%AppData%\planetarium\keystore`
+ */
+export function getDefaultWeb3KeyStorePath(): string {
+  const baseDir =
+    process.platform === "win32"
+      ? process.env.AppData || path.join(homedir(), "AppData", "Roaming")
+      : process.env.XDG_CONFIG_HOME || path.join(homedir(), ".config");
+  // Note that it's not necessary to explicitly choose one of `path.win32` or
+  // `path.posix` here, but it makes unit tests less dependent on mocks:
+  return (process.platform === "win32" ? path.win32 : path.posix).join(
+    baseDir,
+    "planetarium",
+    "keystore",
+  );
+}
+
+const pattern =
+  /^(?:UTC--([0-9]{4}-[0-9]{2}-[0-9]{2})T([0-9]{2}-[0-9]{2}-[0-9]{2})Z--)?([0-9a-f]{8}-(?:[0-9a-f]{4}-){3}[0-9a-f]{12})(?:.json)?$$/i;
+export function parseKeyFilename(
+  name: string,
+): undefined | { keyId: KeyId; createdAt?: Date } {
+  const match = pattern.exec(name);
+  if (match == null) return undefined;
+  return {
+    keyId: match[3].toLowerCase(),
+    createdAt:
+      match[1] != null && match[2] != null
+        ? new Date(`${match[1]}T${match[2].replace(/-/g, ":")}Z`)
+        : undefined,
+  };
+}
+
+export class Web3KeyStore implements ImportableKeyStore<KeyId, RawPrivateKey> {
+  readonly #passphraseEntry: PassphraseEntry;
+
+  readonly path: string;
+
+  constructor(options: Web3KeyStoreOptions) {
+    this.path = options.path ?? getDefaultWeb3KeyStorePath();
+    this.#passphraseEntry = options.passphraseEntry;
+  }
+
+  async *#listKeyFiles(): AsyncIterable<Dirent> {
+    let dir;
+    try {
+      dir = await fs.opendir(this.path);
+    } catch (e) {
+      if (
+        typeof e === "object" &&
+        e != null &&
+        "code" in e &&
+        e.code === "ENOENT"
+      ) {
+        return;
+      }
+      throw e;
+    }
```

Comment:
I guess that this EAFP("It's Easier to Ask Forgiveness than Permission") doesn't seem to help enough, since there are only two cases we concerning; one is ENOENT, another one is just throwing. 🤔

Reply:
The purpose of this code is to return an empty list if the directory containing web3 keystore files is nonexistent(ENOENT), but we wanted to throw when there is another issue preventing from opening the directory.

Reply:
It's just for considering the case where the keystore directory does not exist like it's empty (rather than considering it an exception case).  We left a comment to explain the intention.

Reply:
I understood that intention and thought that using LBYL style(i.e., adding the check for `ENOENT`) without `catch` might be more expressive. 😅

Reply:
Expressiveness-wise, you're right.  However, it's always tricky when we deal with timings… you know.  What if the directory was there a microsecond ago, and then it's suddenly gone when it's starting to read the directory entries?

Reply:
> it's always tricky when we deal with timings… 

🆗  it's also fair enough.